### ISSUE TITLE
Native Assembler: Improvements, Tweaks, Enhancements

### ISSUE THREAD

**Issue by EleanorNB** [ziglang/zig#7561]
Split out from https://github.com/ziglang/zig/issues/2081#issuecomment-751426901 .

# Zig Native Assembler
For system interfacing without libc in LLVM-less builds, we will need our own assembler. This presents us an opportunity to make some improvements. It won't be wise to stray too far, though, as people will need to port their existing code.

## Proposed Changes
- Intel syntax for x86/64, except directives have `.` and symbols have `:`
- [scas](https://github.com/KnightOS/scas/)-style local labels: `.label:`
- Local constants: `=.val`
- `.pub sym`, `.use sym` for sharing symbols within a compilation unit
- `.export sym`, `.extern "file.(o|so)" sym` for sharing symbols between objects
- Declare a symbol as `.pub` or `.export` it inline: `.pub main:`, `.export memcpy:`
- No `.common` or `.global`
- Constants are prepended with `=` everywhere except at declaration, even in `.use` and `.extern`
- `.end` takes a symbol: `.end label` == `.size label, (. - label)`; cannot have both `.end` and `.size`
- Prepend or replace a symbol with its loader address: `0x8000 pin:`, `0xff00:` -- only possible if all previous symbols have been `.end`ed; linker will detect clashes/range errors; local constants are considered to be in scope for the address, but go out of scope at the label
- On the Zig side, `@use(comptime T: type, comptime symbol: []const u8) T` to access `.pub` symbols
- Cull some redundant/historical symbols (flexible)

## Notes
- Local labels are not ambiguous with directives, as directives must begin a line and must not have a colon; a local label will never satisfy both these conditions at once.
- `.pub`/`.use` take advantage of Zig's compilation units: `.pub` symbols are not necessarily exported, and `.export` symbols are not necessarily public.
- `.pub` symbols populate a single global namespace; the amorphous organisation permitted by assembly means a strictly hierarchical symbol-sharing model would be untenable. Explicit `.use` at least makes this much more manageable.
- `.extern "file"` provides some primitive namespacing for libraries, as with Zig `extern` -- without this, making the use of multiple libraries tenable requires a single global namespace for every symbol in every library on the system.
- `.global` makes symbols impossible to track down, `.common` glosses over potential naming errors; their functionality is subsumed by `.pub`/`.use`.
- `.end` was chosen rather than some kind of hierarchical structure or dividing by non-local symbols to allow overlapping of symbols, as well as sequencing:
``` asm
one:
  ; Some code
two:
.end one
  ; This code comes right after `one`, if both are included, but both need not be if optimised
```
This presents an interesting edge case: a local label may be dropped by a non-local symbol while its use would still be valid. I'm not aware of a clean solution to this.
- Prepending a section with a loader address may clash with section declarations in other files, and hence is best left to the build system; also, a specific loader address typically implies specific symbol addresses as well, and we make no guarantees of symbol layout within sections, so it would just be more complication for no reason. This at least gets us a bit closer to #3206.
- I considered a hypothetical `@sImport()`, but the need for strong typing would have made it untenable. Collecting all public symbols into a namespace wouldn't have helped, for the same reason; since builtins will be required anyway, and the lack of explicit source file dependencies means assembly building will have to be coordinated by build.zig anyway, there is no harm in accessing symbols individually. (Note: this still only applies to symbols within the same compilation unit; `.export`ed symbols still come through Zig `extern`, as usual.)

**Comment by pixelherodev** [comment]
Pending the inevitable arguments against it, I'm in full support of this.

I think some of the specific syntax needs to be altered, but aside from that, a "Zig-flavored asm" makes a great deal of sense IMO. This *also* conveniently paves the way for asm files *per input file* instead of per compilation unit (thanks to .use), which allows the assembling pass to be parallelized with a single process per asm file by build.zig instead of needing tracking or dependency code inside of the assembler. This is a great parallel (pun intended) to a Makefile with a bunch of objects built from asm.

I think it might even be able to integrate with incremental compilation; if we have our own syntax, we can just effectively pretend that a .s file is a .zig file with a single `asm` statement.

**Comment by EleanorNB** [comment]
> I think some of the specific syntax needs to be altered,

Which syntax, specifically?

**Comment by pixelherodev** [comment]
@EleanorNB Actually, most of the syntax is fine on a second look, I think.

> .pub sym, .use sym for sharing symbols within a compilation unit

This means multiple input asm files can be compiled into one compilation unit? and `@use` implies that an asm file can become part of a .zig compilation unit, as well?

**Comment by EleanorNB** [comment]
That's right. Zig has multi-file compilation units, we might as well use them. (That's also the only way that `.extern "lib" sym` makes sense, at least until I make my own OS with a custom object format.)

**Comment by pixelherodev** [comment]
If you're serious about making a new object format, it'd probably be easier to add support to an existing OS :P

**Comment by pixelherodev** [comment]
@EleanorNB Would you intend macro support in the asm? If the goal is to be able to write asm directly, I think that makes sense (and a simple macro processor is pretty straightforward).

On the other hand, "macro support" could be a build.zig feature instead, using an extension to build.zig that parses the asm, performs macro replacement, and gives the resultant file as an input instead of the source. This removes the complexity from the compiler, while still allowing asm programmers to add support *themselves* in a manner best suited to their project.

It's worth noting that you could technically do that for Zig code as well, which could theoretically be used to replace some expensive comptime calls with cheap textual substitution. I can already hear people screaming in revulsion at the idea - but the point isn't, "hey, this is a good idea!" it's that it can be done *independently of the compiler!* The advantages of a native build system :)

**Comment by matu3ba** [comment]
@pixelherodev Some examples, how you intent the assembler to look and what parsing+checking synax+type complexity you intend, would be great.

There is a tradeoff between syntax+type complexity = compile-time speed, functional complexity and safety classes.
